### PR TITLE
Add InvalidDefinition and track whether hashes/models failed to setup…

### LINKED ISSUE #127: An exception during attribute block eval leads to impartially-initialized models with improperly setup attributes.
An exception during dsl compiler's `parse` method will result in the class having no attributes defined forever.

### PR DESCRIPTION
**Pull request by careo**
… properly.

Fixes #127

Signed-off-by: Dane Jensen <dane.jensen@gmail.com>

### CODE REVIEW

File: lib/attributor/types/hash.rb [hunk 1]
```diff
@@ -1,4 +1,20 @@
 module Attributor
+  class InvalidDefinition < StandardError
+    def initialize(type, cause)
+      type_name = if type.name
+        type.name
+      else
+        type.inspect
+      end
+
+      msg = "Type #{type_name} is invalid due to exception during definition: #{cause.inspect}."
```
Comment (blanquer) [inline lib/attributor/types/hash.rb]: maybe a more refined message?
"Structure definition for type #{type_name} is invalid. The following exception has occurred: #{cause.inspect}"

File: lib/attributor/types/hash.rb [hunk 2]
```diff
@@ -54,14 +72,20 @@ def self.inherited(klass)
         @value_type = v
         @key_attribute = Attribute.new(@key_type)
         @value_attribute = Attribute.new(@value_type)
+
+        @error = false
```
Comment (blanquer) [inline lib/attributor/types/hash.rb]: do we need this one? if we have it in the top class body?

Comment (careo) [inline lib/attributor/types/hash.rb]: Yes. It's a class instance variable.

**Comment by blanquer** [comment]
LGTM

**Pull request merged**